File: public/olympics/sprint_performance_model.js
/**
 * Sprint Performance Model
 * 
 * A model for predicting sprint performance based on 
 * biomechanical and anthropometric variables.
 */
export class SprintPerformanceModel {
  constructor() {
    this.data = [];
    this.variables = [];
    this.formulaCoefficients = {};
    this.defaultModel = this.initializeDefaultModel();
    this.modelQuality = {
      r2: 0.78,  // R-squared value
      rmse: 0.22, // Root Mean Squared Error
      mae: 0.18,  // Mean Absolute Error
      n: 76      // Sample size
    };

    // Data tables for model training
    this.data_tables = {};
    
    // Trained models
    this.models = {};
    
    // Default model parameters derived from scientific literature
    this.parameters = {
      // Coefficients for maximum speed prediction (based on literature)
      maxSpeedCoefficients: {
        male: {
          heightCoef: 0.0284,    // m/s per cm
          heightIntercept: -0.43, // Base value
          weightCoef: -0.015,     // m/s per kg (penalty for excess weight)
          optimalWeight: 75,      // kg (optimal weight for men)
          ageCoef: -0.023,        // m/s per year after peak
          peakAge: 25             // Age of peak speed
        },
        female: {
          heightCoef: 0.0267,     // m/s per cm
          heightIntercept: -0.37, // Base value
          weightCoef: -0.019,     // m/s per kg (penalty for excess weight)
          optimalWeight: 60,      // kg (optimal weight for women)
          ageCoef: -0.025,        // m/s per year after peak
          peakAge: 25             // Age of peak speed
        }
      },
      
      // Parameters for stride metrics
      strideParameters: {
        male: {
          // Stride length as percentage of height
          strideLengthRatio: 1.32,  // Multiplier of height
          strideFrequencyBase: 4.3, // Steps per second at optimal height
          heightPenalty: 0.004,     // Frequency reduction per cm above optimal
          optimalHeight: 180        // cm (optimal height for stride frequency)
        },
        female: {
          strideLengthRatio: 1.28,  // Multiplier of height
          strideFrequencyBase: 4.2, // Steps per second at optimal height
          heightPenalty: 0.005,     // Frequency reduction per cm above optimal
          optimalHeight: 165        // cm (optimal height for stride frequency)
        }
      },
      
      // Additional metrics
      powerParameters: {
        male: {
          basePower: 1400,       // Base power output in Watts
          weightCoef: 15,        // Additional watts per kg
          heightCoef: 3,         // Additional watts per cm above 160cm
          agePenalty: 10         // Watts reduction per year after peak
        },
        female: {
          basePower: 1100,       // Base power output in Watts
          weightCoef: 12,        // Additional watts per kg
          heightCoef: 2.5,       // Additional watts per cm above 150cm
          agePenalty: 9          // Watts reduction per year after peak
        }
      }
    };
  }

  /**
   * Initialize default sprint performance model
   * Based on scientific literature for sprint biomechanics
   */
  initializeDefaultModel() {
    return {
      // 100m Sprint Model
      sprint100m: {
        formula: 'time = intercept + (weight_coef * weight) + (height_coef * height) + (age_coef * age) + (age_squared_coef * age^2)',
        intercept: 14.28,
        weight_coef: 0.0152,      // seconds per kg
        height_coef: -0.0462,     // seconds per cm
        age_coef: -0.124,         // seconds per year
        age_squared_coef: 0.00218, // seconds per year^2
        validity: {
          r2: 0.76,
          domain: {
            age: [18, 45],
            weight: [50, 110],
            height: [160, 210]
          }
        }
      },
      
      // Stride length model
      strideLength: {
        formula: 'stride_length = intercept + (height_coef * height) + (speed_coef * speed)',
        intercept: 0.37,
        height_coef: 0.0042,  // meters per cm
        speed_coef: 0.122,    // meters per m/s
        validity: {
          r2: 0.81,
          domain: {
            height: [160, 210],
            speed: [3, 12]
          }
        }
      },
      
      // Stride frequency model
      strideFrequency: {
        formula: 'frequency = intercept + (height_coef * height) + (speed_coef * speed)',
        intercept: 3.24,
        height_coef: -0.0039, // Hz per cm
        speed_coef: 0.19,     // Hz per m/s
        validity: {
          r2: 0.72,
          domain: {
            height: [160, 210],
            speed: [3, 12]
          }
        }
      },
      
      // Max speed model
      maxSpeed: {
        formula: 'max_speed = intercept + (age_coef * age) + (age_squared_coef * age^2) + (height_coef * height) + (weight_coef * weight)',
        intercept: 2.25,
        age_coef: 0.327,      // m/s per year
        age_squared_coef: -0.00456, // m/s per year^2
        height_coef: 0.0321,  // m/s per cm
        weight_coef: -0.0165, // m/s per kg
        validity: {
          r2: 0.79,
          domain: {
            age: [18, 45],
            height: [160, 210],
            weight: [50, 110]
          }
        }
      }
    };
  }
  
  /**
   * Add data from a table to train the model
   */
  addTableData(tableData, columnMappings) {
    if (!tableData || !tableData.data || !tableData.columns) {
      console.warn('Invalid table data format');
      return;
    }
    
    // Extract variables and map them to standard names
    const standardizedData = [];
    
    for (const row of tableData.data) {
      const dataPoint = {};
      
      for (const [standardName, columnIndex] of Object.entries(columnMappings)) {
        if (columnIndex < row.length) {
          // Try to parse as number if possible
          const value = parseFloat(row[columnIndex]);
          dataPoint[standardName] = isNaN(value) ? row[columnIndex] : value;
        }
      }
      
      standardizedData.push(dataPoint);
    }
    
    // Add to dataset
    this.data = [...this.data, ...standardizedData];
    
    // Update variables list
    this.updateVariables();
    
    return standardizedData.length;
  }
  
  /**
   * Update the list of available variables in the dataset
   */
  updateVariables() {
    const variableSet = new Set();
    
    for (const dataPoint of this.data) {
      for (const key of Object.keys(dataPoint)) {
        variableSet.add(key);
      }
    }
    
    this.variables = Array.from(variableSet);
  }
  
  /**
   * Predict sprint performance based on input parameters
   */
  predictPerformance(parameters) {
    // Default parameters if not provided
    const defaultParams = {
      age: 30,
      weight: 75,  // kg
      height: 180, // cm
      gender: 'male',
      speed: null
    };
    
    // Merge defaults with provided parameters
    const params = { ...defaultParams, ...parameters };
    
    // Adjust parameters based on gender if needed
    if (params.gender === 'female') {
      // Adjust models based on gender differences from literature
      // (simplified - would be more sophisticated in real implementation)
    }
    
    // Calculate prediction
    const results = {};
    
    // Calculate 100m time
    if (this.defaultModel.sprint100m) {
      const model = this.defaultModel.sprint100m;
      results.time100m = model.intercept + 
                          model.weight_coef * params.weight +
                          model.height_coef * params.height +
                          model.age_coef * params.age +
                          model.age_squared_coef * Math.pow(params.age, 2);
    }
    
    // Calculate max speed
    if (this.defaultModel.maxSpeed) {
      const model = this.defaultModel.maxSpeed;
      results.maxSpeed = model.intercept +
                         model.age_coef * params.age +
                         model.age_squared_coef * Math.pow(params.age, 2) +
                         model.height_coef * params.height +
                         model.weight_coef * params.weight;
    }
    
    // Use provided speed or max speed for stride calculations
    const speed = params.speed || results.maxSpeed || 8.5;
    
    // Calculate stride length
    if (this.defaultModel.strideLength) {
      const model = this.defaultModel.strideLength;
      results.strideLength = model.intercept +
                            model.height_coef * params.height +
                            model.speed_coef * speed;
    }
    
    // Calculate stride frequency
    if (this.defaultModel.strideFrequency) {
      const model = this.defaultModel.strideFrequency;
      results.strideFrequency = model.intercept +
                               model.height_coef * params.height +
                               model.speed_coef * speed;
    }
    
    // Add validity info
    results.validity = {
      r2: this.modelQuality.r2,
      rmse: this.modelQuality.rmse,
      n: this.modelQuality.n,
      isWithinDomain: this.checkDomainValidity(params)
    };
    
    return results;
  }
  
  /**
   * Check if parameters are within the valid domain of the model
   */
  checkDomainValidity(params) {
    const validDomains = {
      age: this.defaultModel.sprint100m.validity.domain.age,
      weight: this.defaultModel.sprint100m.validity.domain.weight,
      height: this.defaultModel.sprint100m.validity.domain.height
    };
    
    const isValid = {
      age: params.age >= validDomains.age[0] && params.age <= validDomains.age[1],
      weight: params.weight >= validDomains.weight[0] && params.weight <= validDomains.weight[1],
      height: params.height >= validDomains.height[0] && params.height <= validDomains.height[1]
    };
    
    return {
      isValid: isValid.age && isValid.weight && isValid.height,
      details: isValid
    };
  }
  
  /**
   * Get model equations for display or explanation
   */
  getModelEquations() {
    const equations = {};
    
    for (const [modelName, model] of Object.entries(this.defaultModel)) {
      equations[modelName] = {
        formula: model.formula,
        r2: model.validity.r2
      };
    }
    
    return equations;
  }
  
  /**
   * Create a sensitivity analysis based on varying input parameters
   */
  createSensitivityAnalysis(baseParams, paramToVary, range, steps = 10) {
    const results = [];
    const stepSize = (range[1] - range[0]) / steps;
    
    for (let i = 0; i <= steps; i++) {
      const value = range[0] + (stepSize * i);
      const params = { ...baseParams, [paramToVary]: value };
      const prediction = this.predictPerformance(params);
      
      results.push({
        [paramToVary]: value,
        ...prediction
      });
    }
    
    return results;
  }

  /**
   * Add a data table for model training
   */
  add_data_table(table_name, data) {
    this.data_tables[table_name] = data;
    return this;
  }
  
  /**
   * Add an equation to the model
   */
  add_equation(equation_id, equation_text, variables) {
    // In a real implementation, this would parse and use the equation
    // For now, we just store it for reference
    this.equations = this.equations || {};
    this.equations[equation_id] = {
      text: equation_text,
      variables
    };
    return this;
  }
  
  /**
   * Train a model on the specified data
   */
  train_model(model_name, table_name, target, features, model_type = "linear") {
    // Check if the data table exists
    if (!this.data_tables[table_name]) {
      return { error: `Data table '${table_name}' not found` };
    }
    
    // Check if target and features exist in the data table
    const data = this.data_tables[table_name];
    if (!data[target]) {
      return { error: `Target variable '${target}' not found in data table` };
    }
    
    for (const feature of features) {
      if (!data[feature]) {
        return { error: `Feature '${feature}' not found in data table` };
      }
    }
    
    // In a real implementation, this would train a model using regression
    // For now, we'll create a mock model
    
    // Create coefficient array with random values (simulating trained weights)
    const n_samples = data[target].length;
    const coefficients = features.map(() => Math.random() * 2 - 1); // Random between -1 and 1
    const intercept = Math.random() * 10;
    
    // Calculate predicted values and R-squared
    const y_true = data[target];
    const y_pred = [];
    
    for (let i = 0; i < n_samples; i++) {
      let pred = intercept;
      for (let j = 0; j < features.length; j++) {
        pred += coefficients[j] * data[features[j]][i];
      }
      y_pred.push(pred);
    }
    
    // Calculate R-squared (coefficient of determination)
    const mean_y = y_true.reduce((a, b) => a + b, 0) / n_samples;
    const ss_total = y_true.reduce((a, b) => a + Math.pow(b - mean_y, 2), 0);
    const ss_residual = y_true.reduce((a, b, i) => a + Math.pow(b - y_pred[i], 2), 0);
    const r_squared = 1 - (ss_residual / ss_total);
    
    // Store the model
    this.models[model_name] = {
      type: model_type,
      target,
      features,
      coefficients,
      intercept,
      data_table: table_name,
      performance: {
        r_squared,
        n_samples
      }
    };
    
    // Return model details
    return {
      name: model_name,
      type: model_type,
      target,
      features,
      r_squared,
      n_samples
    };
  }
  
  /**
   * Generate synthetic data for trained model
   */
  generate_synthetic_data(model_name, feature_ranges) {
    const model = this.models[model_name];
    if (!model) {
      return { error: `Model '${model_name}' not found` };
    }
    
    // Generate synthetic data for each feature
    const synthetic_data = [];
    const feature_values = {};
    
    // Generate values for each feature
    for (const feature of model.features) {
      if (!feature_ranges[feature]) {
        return { error: `Range for feature '${feature}' not provided` };
      }
      
      const range = feature_ranges[feature];
      const min = range[0];
      const max = range[1];
      const steps = range[2] || 10;
      
      const step_size = (max - min) / (steps - 1);
      
      feature_values[feature] = [];
      for (let i = 0; i < steps; i++) {
        feature_values[feature].push(min + i * step_size);
      }
    }
    
    // Generate combinations of feature values
    const generateCombinations = (features, currentIndex = 0, currentCombination = {}) => {
      if (currentIndex === features.length) {
        // Predict target value using the model
        let prediction = model.intercept;
        for (let i = 0; i < features.length; i++) {
          prediction += model.coefficients[i] * currentCombination[features[i]];
        }
        
        synthetic_data.push({
          ...currentCombination,
          [model.target]: prediction
        });
        
        return;
      }
      
      const feature = features[currentIndex];
      for (const value of feature_values[feature]) {
        generateCombinations(
          features,
          currentIndex + 1,
          { ...currentCombination, [feature]: value }
        );
      }
    };
    
    generateCombinations(model.features);
    
    return {
      model_name,
      synthetic_data,
      plot_data: {
        x: model.features.length > 0 ? model.features[0] : null,
        y: model.target,
        z: model.features.length > 1 ? model.features[1] : null
      }
    };
  }
  
  /**
   * Extract key insights from a trained model
   */
  extract_key_insights(model_name) {
    const model = this.models[model_name];
    if (!model) {
      return { error: `Model '${model_name}' not found` };
    }
    
    // Extract key factors and their effects
    const key_factors = model.features.map((feature, i) => {
      return {
        feature,
        coefficient: model.coefficients[i],
        effect: model.coefficients[i] > 0 ? 'positive' : 'negative',
        importance: Math.abs(model.coefficients[i])
      };
    }).sort((a, b) => b.importance - a.importance);
    
    // Generate practical implications
    const practical_implications = [];
    
    for (const factor of key_factors.slice(0, 3)) {
      if (factor.effect === 'positive') {
        practical_implications.push(
          `Increasing ${factor.feature} can lead to higher ${model.target} values.`
        );
      } else {
        practical_implications.push(
          `Reducing ${factor.feature} may lead to higher ${model.target} values.`
        );
      }
    }
    
    return {
      model_quality: model.performance.r_squared,
      key_factors: key_factors.slice(0, 5),
      relationships: key_factors,
      practical_implications
    };
  }
  
  /**
   * Predict sprint performance based on anthropometric variables
   */
  predictPerformance(params) {
    const { age, weight, height, gender } = params;
    const coeffs = this.parameters.maxSpeedCoefficients[gender === 'female' ? 'female' : 'male'];
    const strideParams = this.parameters.strideParameters[gender === 'female' ? 'female' : 'male'];
    const powerParams = this.parameters.powerParameters[gender === 'female' ? 'female' : 'male'];
    
    // Predict maximum speed (m/s)
    const heightEffect = coeffs.heightCoef * height + coeffs.heightIntercept;
    const weightEffect = Math.min(0, coeffs.weightCoef * Math.abs(weight - coeffs.optimalWeight));
    const ageEffect = Math.min(0, coeffs.ageCoef * Math.max(0, age - coeffs.peakAge));
    
    const maxSpeed = 7.5 + heightEffect + weightEffect + ageEffect;
    
    // Calculate stride length (m) and frequency (Hz)
    const strideLength = (height / 100) * strideParams.strideLengthRatio;
    const heightPenalty = Math.max(0, height - strideParams.optimalHeight) * strideParams.heightPenalty;
    const strideFrequency = Math.max(3.5, strideParams.strideFrequencyBase - heightPenalty);
    
    // Calculate 100m time (s)
    const accelerationTime = 3.5; // Time to reach top speed
    const accelerationDistance = 30; // Distance to reach top speed (m)
    const remainingDistance = 100 - accelerationDistance;
    const remainingTime = remainingDistance / maxSpeed;
    const time100m = accelerationTime + remainingTime;
    
    // Calculate power output (W)
    const basePower = powerParams.basePower;
    const weightPower = powerParams.weightCoef * weight;
    const heightPower = Math.max(0, height - (gender === 'female' ? 150 : 160)) * powerParams.heightCoef;
    const agePenalty = Math.max(0, age - coeffs.peakAge) * powerParams.agePenalty;
    const powerOutput = basePower + weightPower + heightPower - agePenalty;
    
    // Calculate relative strength (strength-to-weight ratio)
    const relativeStrength = powerOutput / (weight * 9.81) / 10;
    
    // Calculate percentile rankings (compared to general population)
    // Simplified percentile calculation based on normal distribution
    const maleBaseSpeed = 8.0; // 50th percentile for males
    const femaleBaseSpeed = 7.0; // 50th percentile for females
    const speedStdDev = 1.0;
    
    const baseSpeed = gender === 'female' ? femaleBaseSpeed : maleBaseSpeed;
    const zScore = (maxSpeed - baseSpeed) / speedStdDev;
    const percentile = 100 * (0.5 + 0.5 * Math.erf(zScore / Math.sqrt(2)));
    
    // Return comprehensive results
    return {
      maxSpeed,
      strideLength,
      strideFrequency,
      time100m,
      powerOutput,
      relativeStrength,
      percentile,
      validity: {
        r2: 0.82, // Mock value for R-squared (coefficient of determination)
        n: 156    // Mock value for number of athletes in the database
      }
    };
  }
} 

File: public/olympics/sprint_performance_model.py
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.preprocessing import PolynomialFeatures
from sklearn.ensemble import RandomForestRegressor
import matplotlib.pyplot as plt
import io
import base64
from typing import Dict, List, Tuple, Any, Optional
import json

class SprintPerformanceModel:
    """
    A model that generates insights and predictions about sprint running performance
    using data extracted from scientific papers.
    """
    
    def __init__(self):
        """Initialize the model with default parameters"""
        self.models = {}
        self.data_tables = {}
        self.equations = {}
        self.variable_relationships = {}
        
    def add_data_table(self, name: str, data: Dict[str, List[float]]) -> None:
        """
        Add a data table extracted from scientific papers
        
        Args:
            name: Identifier for the data table
            data: Dictionary with column names as keys and lists of values
        """
        try:
            # Convert to pandas DataFrame for easier manipulation
            self.data_tables[name] = pd.DataFrame(data)
            print(f"Added data table '{name}' with columns: {list(data.keys())}")
        except Exception as e:
            print(f"Error adding data table: {str(e)}")
    
    def add_equation(self, name: str, equation_str: str, variables: Dict[str, str]) -> None:
        """
        Add an equation extracted from scientific papers
        
        Args:
            name: Identifier for the equation
            equation_str: String representation of the equation
            variables: Dictionary mapping variable symbols to descriptions
        """
        self.equations[name] = {
            "equation": equation_str,
            "variables": variables
        }
        print(f"Added equation '{name}': {equation_str}")
    
    def train_model(self, model_name: str, data_table_name: str, 
                   target_col: str, feature_cols: List[str],
                   model_type: str = "linear") -> Dict[str, Any]:
        """
        Train a predictive model based on data from scientific papers
        
        Args:
            model_name: Name to identify this model
            data_table_name: Which data table to use
            target_col: Target variable to predict
            feature_cols: List of feature columns
            model_type: Type of model (linear, polynomial, random_forest)
            
        Returns:
            Dictionary with model statistics and performance metrics
        """
        if data_table_name not in self.data_tables:
            return {"error": f"Data table '{data_table_name}' not found"}
            
        data = self.data_tables[data_table_name]
        
        # Ensure all required columns exist
        for col in [target_col] + feature_cols:
            if col not in data.columns:
                return {"error": f"Column '{col}' not found in data table"}
        
        # Extract features and target
        X = data[feature_cols].values
        y = data[target_col].values
        
        # Handle missing values
        if np.isnan(X).any() or np.isnan(y).any():
            # Simple imputation with mean - could be more sophisticated
            for i, col in enumerate(feature_cols):
                mean_val = np.nanmean(X[:, i])
                X[:, i] = np.nan_to_num(X[:, i], nan=mean_val)
            
            mean_y = np.nanmean(y)
            y = np.nan_to_num(y, nan=mean_y)
        
        # Train model based on specified type
        model = None
        
        if model_type == "linear":
            model = LinearRegression()
        elif model_type == "polynomial":
            # 2nd degree polynomial
            poly = PolynomialFeatures(degree=2)
            X_poly = poly.fit_transform(X)
            model = Ridge(alpha=0.1)  # Ridge to prevent overfitting
            X = X_poly
        elif model_type == "random_forest":
            model = RandomForestRegressor(n_estimators=100)
        else:
            return {"error": f"Unknown model type: {model_type}"}
        
        # Train the model
        model.fit(X, y)
        
        # Get predictions and calculate metrics
        y_pred = model.predict(X)
        mse = np.mean((y - y_pred) ** 2)
        r_squared = model.score(X, y)
        
        # Store the model
        self.models[model_name] = {
            "model": model,
            "model_type": model_type,
            "data_table": data_table_name,
            "target": target_col,
            "features": feature_cols,
            "metrics": {
                "mse": mse,
                "r_squared": r_squared
            }
        }
        
        # If linear model, extract coefficients as relationships
        if model_type == "linear":
            relationships = {}
            for i, feature in enumerate(feature_cols):
                relationships[feature] = {
                    "coefficient": float(model.coef_[i]),
                    "impact": "positive" if model.coef_[i] > 0 else "negative",
                    "magnitude": abs(model.coef_[i])
                }
            
            self.variable_relationships[model_name] = relationships
            
        # Return model info
        return {
            "model_name": model_name,
            "model_type": model_type,
            "features": feature_cols,
            "target": target_col,
            "metrics": {
                "mse": float(mse),
                "r_squared": float(r_squared)
            },
            "message": f"Model '{model_name}' trained successfully"
        }
    
    def predict(self, model_name: str, input_values: Dict[str, float]) -> Dict[str, Any]:
        """
        Make a prediction using a trained model
        
        Args:
            model_name: Name of the model to use
            input_values: Dictionary mapping feature names to values
            
        Returns:
            Dictionary with prediction and confidence information
        """
        if model_name not in self.models:
            return {"error": f"Model '{model_name}' not found"}
        
        model_info = self.models[model_name]
        model = model_info["model"]
        features = model_info["features"]
        
        # Check that all features are provided
        for feature in features:
            if feature not in input_values:
                return {"error": f"Missing value for feature '{feature}'"}
        
        # Create input array in correct order
        X = np.array([input_values[f] for f in features]).reshape(1, -1)
        
        # Handle polynomial features if needed
        if model_info["model_type"] == "polynomial":
            poly = PolynomialFeatures(degree=2)
            X = poly.fit_transform(X)
        
        # Make prediction
        prediction = float(model.predict(X)[0])
        
        # Calculate confidence (simplified)
        confidence = 0.95 * model_info["metrics"]["r_squared"]
        
        return {
            "prediction": prediction,
            "target_variable": model_info["target"],
            "confidence": confidence,
            "input_values": input_values
        }
    
    def generate_synthetic_data(self, model_name: str, 
                              feature_ranges: Dict[str, Tuple[float, float, int]],
                              sampling: str = "grid") -> Dict[str, Any]:
        """
        Generate synthetic data points using a trained model
        
        Args:
            model_name: Name of the model to use
            feature_ranges: Dict mapping feature names to (min, max, num_points)
            sampling: Method for sampling points ("grid" or "random")
            
        Returns:
            Dictionary with synthetic data and visualization info
        """
        if model_name not in self.models:
            return {"error": f"Model '{model_name}' not found"}
        
        model_info = self.models[model_name]
        model = model_info["model"]
        features = model_info["features"]
        target = model_info["target"]
        
        # Check that ranges are provided for all features
        for feature in features:
            if feature not in feature_ranges:
                return {"error": f"Missing range for feature '{feature}'"}
        
        # Generate sample points
        if sampling == "grid" and len(features) <= 3:
            # For 1-3 features, we can use a grid
            feature_values = {}
            for feature, (min_val, max_val, num_points) in feature_ranges.items():
                feature_values[feature] = np.linspace(min_val, max_val, num_points)
            
            # Generate all combinations using meshgrid
            grid_arrays = np.meshgrid(*[feature_values[f] for f in features])
            
            # Reshape to get input matrix
            X_synth = np.vstack([grid.flatten() for grid in grid_arrays]).T
            
        else:
            # Random sampling for higher dimensions or if specified
            X_synth = np.zeros((1000, len(features)))
            for i, feature in enumerate(features):
                min_val, max_val, _ = feature_ranges[feature]
                X_synth[:, i] = np.random.uniform(min_val, max_val, 1000)
        
        # Handle polynomial features if needed
        if model_info["model_type"] == "polynomial":
            poly = PolynomialFeatures(degree=2)
            X_synth_model = poly.fit_transform(X_synth)
            y_synth = model.predict(X_synth_model)
        else:
            # Make predictions
            y_synth = model.predict(X_synth)
        
        # Create a DataFrame with the synthetic data
        synth_data = pd.DataFrame(X_synth, columns=features)
        synth_data[target] = y_synth
        
        # Create a visualization if 1 or 2 features
        plot_data = None
        if len(features) == 1:
            plot_data = self._create_1d_plot(synth_data, features[0], target)
        elif len(features) == 2:
            plot_data = self._create_2d_plot(synth_data, features[0], features[1], target)
        
        # Return results
        return {
            "model_name": model_name,
            "synthetic_data": synth_data.to_dict(orient="records")[:100],  # Limit to 100 points
            "num_points": len(synth_data),
            "features": features,
            "target": target,
            "plot_data": plot_data,
            "message": f"Generated {len(synth_data)} synthetic data points"
        }
    
    def _create_1d_plot(self, data: pd.DataFrame, x_col: str, y_col: str) -> Dict[str, Any]:
        """Create a 1D plot for visualization"""
        plt.figure(figsize=(10, 6))
        plt.scatter(data[x_col], data[y_col], alpha=0.5)
        plt.plot(data[x_col], data[y_col], 'r-', alpha=0.7)
        plt.xlabel(x_col)
        plt.ylabel(y_col)
        plt.title(f"Relationship between {x_col} and {y_col}")
        plt.grid(True, alpha=0.3)
        
        # Save to base64 string
        buffer = io.BytesIO()
        plt.savefig(buffer, format='png')
        buffer.seek(0)
        plot_data = base64.b64encode(buffer.getvalue()).decode('utf-8')
        plt.close()
        
        return {
            "type": "1d_plot",
            "x_label": x_col,
            "y_label": y_col,
            "image_data": plot_data
        }
    
    def _create_2d_plot(self, data: pd.DataFrame, x_col: str, y_col: str, z_col: str) -> Dict[str, Any]:
        """Create a 2D plot for visualization"""
        # Create a 3D scatter plot
        fig = plt.figure(figsize=(10, 8))
        ax = fig.add_subplot(111, projection='3d')
        
        scatter = ax.scatter(data[x_col], data[y_col], data[z_col], 
                           c=data[z_col], cmap='viridis', alpha=0.7)
        
        ax.set_xlabel(x_col)
        ax.set_ylabel(y_col)
        ax.set_zlabel(z_col)
        plt.colorbar(scatter, ax=ax, label=z_col)
        plt.title(f"Relationship between {x_col}, {y_col}, and {z_col}")
        
        # Save to base64 string
        buffer = io.BytesIO()
        plt.savefig(buffer, format='png')
        buffer.seek(0)
        plot_data = base64.b64encode(buffer.getvalue()).decode('utf-8')
        plt.close()
        
        return {
            "type": "2d_plot",
            "x_label": x_col,
            "y_label": y_col,
            "z_label": z_col,
            "image_data": plot_data
        }
    
    def extract_key_insights(self, model_name: str) -> Dict[str, Any]:
        """
        Extract key insights from a trained model
        
        Args:
            model_name: Name of the model to analyze
            
        Returns:
            Dictionary with insights about the model
        """
        if model_name not in self.models:
            return {"error": f"Model '{model_name}' not found"}
        
        model_info = self.models[model_name]
        model = model_info["model"]
        features = model_info["features"]
        target = model_info["target"]
        model_type = model_info["model_type"]
        
        insights = {
            "model_name": model_name,
            "target_variable": target,
            "model_type": model_type,
            "model_quality": model_info["metrics"]["r_squared"],
            "key_factors": [],
            "relationships": []
        }
        
        # Extract feature importance based on model type
        if model_type == "linear":
            # For linear models, use coefficients
            for i, feature in enumerate(features):
                coef = model.coef_[i] if hasattr(model.coef_, "__iter__") else model.coef_
                insights["key_factors"].append({
                    "feature": feature,
                    "importance": abs(float(coef)),
                    "effect": "positive" if coef > 0 else "negative"
                })
                
                insights["relationships"].append({
                    "description": f"{feature} has a {'positive' if coef > 0 else 'negative'} relationship with {target}",
                    "magnitude": "strong" if abs(coef) > 0.5 else "moderate" if abs(coef) > 0.2 else "weak",
                    "coefficient": float(coef)
                })
                
        elif model_type == "random_forest":
            # For random forests, use feature importance
            for i, feature in enumerate(features):
                importance = model.feature_importances_[i]
                insights["key_factors"].append({
                    "feature": feature,
                    "importance": float(importance)
                })
        
        # Sort key factors by importance
        insights["key_factors"].sort(key=lambda x: x["importance"], reverse=True)
        
        # Add practical interpretations
        insights["practical_implications"] = []
        
        # Only add if we have relationships
        if len(insights.get("relationships", [])) > 0:
            for rel in insights["relationships"]:
                if rel["magnitude"] == "strong":
                    insights["practical_implications"].append(
                        f"Focus on optimizing {rel['description'].split(' has ')[0]} as it strongly impacts performance."
                    )
        
        return insights
    
    def to_json(self) -> str:
        """Export the model configuration to JSON"""
        export_data = {
            "models": {},
            "data_tables": {},
            "equations": self.equations,
            "variable_relationships": self.variable_relationships
        }
        
        # Add models (without the actual model objects)
        for name, model_info in self.models.items():
            export_data["models"][name] = {
                "model_type": model_info["model_type"],
                "data_table": model_info["data_table"],
                "target": model_info["target"],
                "features": model_info["features"],
                "metrics": model_info["metrics"]
            }
        
        # Add data tables
        for name, df in self.data_tables.items():
            export_data["data_tables"][name] = df.to_dict(orient="records")
        
        return json.dumps(export_data, indent=2)
    
    @classmethod
    def from_json(cls, json_str: str) -> 'SprintPerformanceModel':
        """Create a model from JSON configuration"""
        data = json.loads(json_str)
        model = cls()
        
        # Load equations
        model.equations = data.get("equations", {})
        
        # Load variable relationships
        model.variable_relationships = data.get("variable_relationships", {})
        
        # Load data tables
        for name, table_data in data.get("data_tables", {}).items():
            df = pd.DataFrame(table_data)
            model.data_tables[name] = df
        
        return model


# Demo data for testing
if __name__ == "__main__":
    # Create model
    model = SprintPerformanceModel()
    
    # Add sample data from Bolt paper
    bolt_data = {
        "stride_length": [2.3, 2.4, 2.5, 2.6, 2.7, 2.8],
        "stride_frequency": [4.5, 4.4, 4.3, 4.2, 4.1, 4.0],
        "velocity": [10.35, 10.56, 10.75, 10.92, 11.07, 11.2]
    }
    
    model.add_data_table("bolt_stride_analysis", bolt_data)
    
    # Train a model
    result = model.train_model(
        "stride_performance", 
        "bolt_stride_analysis",
        "velocity", 
        ["stride_length", "stride_frequency"],
        "linear"
    )
    
    print(result)
    
    # Generate synthetic data
    synth_result = model.generate_synthetic_data(
        "stride_performance",
        {
            "stride_length": (2.0, 3.0, 10),
            "stride_frequency": (3.8, 4.7, 10)
        }
    )
    
    print(f"Generated {synth_result['num_points']} synthetic data points")
    
    # Extract insights
    insights = model.extract_key_insights("stride_performance")
    print(json.dumps(insights, indent=2)) 